### PR TITLE
[ID-1120] Add /api/serviceaccount/v1/{provider}/key endpoint

### PR DESCRIPTION
**Pull request by samanehsan**
**Jira:** https://broadworkbench.atlassian.net/browse/ID-1120

**What:** 
Add an `/api/serviceaccount/v1/{provider}/key` endpoint that uses the requesting user's id and the provider to look up a fenceAccountKey from the database, and return it if it exists and is not expired.

**Why:**
Identiteam is getting rid of Bond and adding its functionality to ECM to consolidate these similar services. This new endpoint is equivalent to the Bond `/api/link/v1/{provider}/serviceaccount/key` [endpoint](https://bond.dsde-dev.broadinstitute.org/api/docs/#/).

### CODE REVIEW

File: common/openapi.yml [hunk 1]
```diff
@@ -223,6 +223,25 @@ paths:
         '500':
           $ref: '#/components/responses/ServerError'
 
+  /api/fenceAccountKey/v1/{provider}/key:
```
Comment (tlangs) [inline common/openapi.yml]: if the api group is `fenceAccountKey`, do we really need the `/key` on the end?

File: service/src/main/java/bio/terra/externalcreds/controllers/FenceAccountKeyController.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+  @Override
+  public ResponseEntity<String> getFenceAccountKey(Provider provider) {
+    var samUser = samUserFactory.from(request);
+    var maybeLinkedAccount =
+        linkedAccountService.getLinkedAccount(samUser.getSubjectId(), provider);
+    var maybeFenceAccountKey =
+        fenceAccountKeyService.getFenceAccountKey(samUser.getSubjectId(), provider);
+
+    auditLogger.logEvent(
```
Comment (tlangs) [inline service/src/main/java/bio/terra/externalcreds/controllers/FenceAccountKeyController.java]: If the linked account doesn't exist, wouldn't we want to just return a 404? We definitely don't want to log an event if there's no linked account.

File: service/src/main/java/bio/terra/externalcreds/controllers/FenceAccountKeyController.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    LinkedAccountService linkedAccountService)
+    implements FenceAccountKeyApi {
+
+  @Override
+  public ResponseEntity<String> getFenceAccountKey(Provider provider) {
+    var samUser = samUserFactory.from(request);
+    var maybeLinkedAccount =
+        linkedAccountService.getLinkedAccount(samUser.getSubjectId(), provider);
+    var maybeFenceAccountKey =
+        fenceAccountKeyService.getFenceAccountKey(samUser.getSubjectId(), provider);
```
Comment (tlangs) [inline service/src/main/java/bio/terra/externalcreds/controllers/FenceAccountKeyController.java]: We should probably just `flatMap` the linked account. We could get the userId and provider from the linked account if it exists.

**Review by tlangs — CHANGES_REQUESTED** [review]

File: common/openapi.yml (hunk 1 above)
Comment (samanehsan) [reply common/openapi.yml]: That's a good point! I agree it's redundant

File: service/src/main/java/bio/terra/externalcreds/controllers/FenceAccountKeyController.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    HttpServletRequest request,
+    LinkedAccountService linkedAccountService)
+    implements FenceAccountKeyApi {
+
+  @Override
+  public ResponseEntity<String> getFenceAccountKey(Provider provider) {
+    var samUser = samUserFactory.from(request);
+    var maybeLinkedAccount =
+        linkedAccountService.getLinkedAccount(samUser.getSubjectId(), provider);
+    var response =
```
Comment (tlangs) [inline service/src/main/java/bio/terra/externalcreds/controllers/FenceAccountKeyController.java]: Could we de-nest these `flatMap`s? It gets hard to read when we start nesting. Saving each intermediate `Optional` also improves debuggability. To do that, we can just declare an `AuditLogEvent.Builder` at the top of the method, and then fill in appropriate details as we go, finally `build()`ing and logging the event as we craft the response.

**Comment by samanehsan** [comment]
The tests are failing because the fence providers are not defined in the DB enum. That will get added in https://github.com/DataBiosphere/terra-external-credentials-manager/pull/178!

**Review by Shakespeared — APPROVED** [review]

**Review by tlangs — APPROVED** [review]

**Review by Ghost-in-a-Jar — APPROVED** [review]

**Pull request merged**